logger/analytics/analyticslogger.go

```diff
@@ -89,22 +127,69 @@ func (al *Logger) Write(bs []byte) (int, error) {
 		return 0, err
 	}
 	bs = append(bs, '\n')
-	r := retrier.New(retrier.ExponentialBackoff(5, 100*time.Millisecond), RequestErrorClassifier{})
-	if err := r.Run(func() error {
-		_, err := al.fhAPI.PutRecord(&firehose.PutRecordInput{
-			DeliveryStreamName: aws.String(al.fhStream),
-			Record:             &firehose.Record{Data: bs},
-		})
-		if err != nil {
-			return err
-		}
-		return nil
-	}); err != nil {
-		return 0, fmt.Errorf("PutRecords: %v", err)
+	al.mu.Lock()
+	al.batchBytes += len(bs)
+	al.batch = append(al.batch, &firehose.Record{Data: bs})
+	shouldSendBatch := len(al.batch) == al.maxBatchRecords || al.batchBytes > int(0.9*float64(al.maxBatchBytes))
+	if shouldSendBatch {
+		batch := al.batch
+		al.batch = []*firehose.Record{}
+		al.batchBytes = 0
+		// be careful not to send al.batch, since we will unlock before we finish sending the batch
+		go sendBatch(batch, al.fhAPI, al.fhStream, time.Now().Add(timeoutForSendingBatches))
```

Comment:
This seems like a reasonable approach! My main concern is a consumer of this library never receives any feedback on errors writing throughout program execution. It feels like a relatively large hole considering the spectrum of errors that can occur (e.g. the user has incorrectly setup perms so every single write fails). As a user I would be confused if I never received an error

Reply:
adding configurable logger to use for errors seems reasonable, maybe defaulting to something simple

Reply:
another option might be to persist an error on the `al` and return it on the next `Write`. I'm honestly totally sure what pattern to apply

Reply:
I don't think that will work, since underneath the hood kayvee `Logger`s use a https://golang.org/pkg/log `Logger`, specifically the `SetOutput` + `Println` methods. The `Println` method within `pkg/log` is the one that eventually triggers the call to `Write`, but it ignores the error. There is an alternative to `Println` called `pkg/log.Logger` called `Output` that returns the error, but then the problem becomes how should all kayvee loggers treat the error returned.